test/recipes/70-test_sslvertol.t

```diff
@@ -34,33 +34,71 @@ my $proxy = TLSProxy::Proxy->new(
     (!$ENV{HARNESS_ACTIVE} || $ENV{HARNESS_VERBOSE})
 );
 
+my @available_tls_versions = ();
+foreach (available_protocols("tls")) {
+    unless (disabled($_)) {
+        note("Checking enabled protocol $_");
+        m|^([a-z]+)(\d)(_\d)?|;
+        my $versionname;
+        if (defined $3) {
+            $versionname = 'TLSProxy::Record::VERS_'.uc($1).'_'.$2.$3;
+            note("'$1', '$2', '$3' => $versionname");
+        } else {
+            $versionname = 'TLSProxy::Record::VERS_'.uc($1).'_'.$2.'_0';
+            note("'$1', '$2' => $versionname");
+        }
+        push @available_tls_versions, eval $versionname;
+    }
+}
+note("TLS versions we can expect: ", join(", ", @available_tls_versions));
+
 #This file does tests without the supported_versions extension.
 #See 70-test_sslversions.t for tests with supported versions.
-#Test 1: Asking for TLS1.4 should pass and negotiate TLS1.2
+#Test 1: Asking for TLS1.4 should pass and negotiate the maximum
+#available TLS version according to configuration
 my $client_version = TLSProxy::Record::VERS_TLS_1_4;
-$proxy->clientflags("-no_tls1_3");
 $proxy->start() or plan skip_all => "Unable to start up Proxy for tests";
-plan tests => 3;
+plan tests => 4;
 my $record = pop @{$proxy->record_list};
-ok(TLSProxy::Message->success()
-   && $record->version() == TLSProxy::Record::VERS_TLS_1_2,
-   "Version tolerance test, TLS 1.4");
+ok((note("Record version received: ".$record->version()),
+    TLSProxy::Message->success())
+   && $record->version() == tls_version_below(TLSProxy::Record::VERS_TLS_1_4),
+   "Version tolerance test, below TLS 1.4");
 
-#Test 2: Asking for TLS1.3 should succeed and negotiate TLS1.2
+#Test 2: Asking for TLS1.3 with that disabled should succeed and negotiate
+#the highest configured TLS version below that.
+$client_version = TLSProxy::Record::VERS_TLS_1_3;
 $proxy->clear();
 $proxy->clientflags("-no_tls1_3");
 $proxy->start();
 $record = pop @{$proxy->record_list};
-ok(TLSProxy::Message->success()
-   && $record->version() == TLSProxy::Record::VERS_TLS_1_2,
-   "Version tolerance test, TLS 1.3");
+ok((note("Record version received: ".$record->version()),
+    TLSProxy::Message->success())
+   && $record->version() == tls_version_below(TLSProxy::Record::VERS_TLS_1_3),
+   "Version tolerance test, max version but not TLS 1.3");
 
-#Test 3: Testing something below SSLv3 should fail
+#Test 3: Asking for TLS1.3 should succeed and negotiate TLS1.3, or if
+#it's disabled, the max version below that.
+$client_version = TLSProxy::Record::VERS_TLS_1_3;
+$proxy->clear();
+$proxy->start();
+$record = pop @{$proxy->record_list};
+ok((note("Record version received: ".$record->version()),
+    TLSProxy::Message->success())
+   && $record->version() == tls_version_below(TLSProxy::Record::VERS_TLS_1_4),
+   "Version tolerance test, anything up to and including TLS 1.3");
```

Comment:
`supported_versions` had escaped me.  I'll simply remove this test.

As for line 55, I'm not sure how that's relevant.  That only tells which protocols are available per OpenSSL configuration, nothing else.  The tests may then apply their own conditions.  Basically, I needed line 55 so it was clear to me exactly what protocol numbers I can expect at any time.